src/BooksDisplayer/BooksDisplayer.tsx

```diff
@@ -1,189 +1,236 @@
-      setFilteredBooks(filteredBooks);
-      setExpandedBook(null);
-   }, [name, records]);
-   
-   function filterByCategories(records, category){
+  const [records, setRecords] = useState<Book[]>([]);
+  const [expandedBook, setExpandedBook] = useState<number | null>(null);
+  const [filteredBooks, setFilteredBooks] = useState<Book[]>([]);
+  const [favorites, setFavorites] = useState<Book[]>([]);
+  const [category, setCategory] = useState("All");
+  const [name, setName] = useState("");
+  const [isLoading, setIsLoading] = useState(true);
+
+  interface Book {
+    id: number;
+    title: string;
+    authors: Array<{ name: string; birth_year?: number; death_year?: number }>;
+    subjects: string[];
+    bookshelves: string[];
+    copyright: boolean;
+    download_count: number;
+    formats: { [format: string]: string };
+    languages: string[];
+    media_type: string;
+    translators: unknown[];
+  }
+  // Fetching data from API
+  useEffect(() => {
+    setIsLoading(true);
+    fetch("https://gutendex.com/books")
+      .then((response) => response.json())
+      .then((data: { results: Book[] }) => {
+        setRecords(data.results);
+        setIsLoading(false);
+      })
+      .catch((error) => {
+        setIsLoading(false);
+        throw new Error(error as string);
+      });
+  }, []);
+  //Filtering books by categories/name
+  useEffect(() => {
+    function filterByCategories() {
       if (category === "All") {
-         return records;
-     }
-
-     if (category === "Favorite"){
-      return favorites;
-     }
- 
-     const regex = new RegExp(category);
- 
-     const filteredBooks = records.filter(book => {
-         const subjects = book.subjects || [];
-         return subjects.some(subject => regex.test(subject));
-     });
- 
-     return filteredBooks;
-   }
-
-
-   function filterByName(records, name){
+        return records;
+      }
+      if (category === "Favorite") {
+        return favorites;
+      }
+      const regex = new RegExp(category);
+
+      const outputFilteredBooks = records.filter((book) => {
+        const subjects = book.subjects;
+        return subjects.some((subject: string) => regex.test(subject));
+      });
+      return outputFilteredBooks;
+    }
+    setFilteredBooks(filterByCategories());
+    setExpandedBook(null);
+  }, [category, records, favorites]);
+
+  useEffect(() => {
+    function filterByName() {
       const nameToLowerCase = name.toLowerCase();
-      const filteredBooks = records.filter(book =>{
-         const title = book.title.toLowerCase();
-         return title.includes(nameToLowerCase);
+      const outputFilteredBooks = records.filter((book: { title: string }) => {
+        const title = book.title.toLowerCase();
+        return title.includes(nameToLowerCase);
       });
 
-      return filteredBooks;
-   }
+      return outputFilteredBooks;
+    }
+    setFilteredBooks(filterByName());
+    setExpandedBook(null);
+  }, [records, name]);
```

Comment:
uważaj bardzo na używanie useEffect'ów, powinno być ich jak najmniej w kodzie, bo bardzo komplikują całe flow

to jest akurat złe użycie useEffecta, bo on służy do synchronizacji z zewnętrznymi systemami (jakieś api, local storage, kamerka), a tutaj używasz samego api reacta, czyli mogłeś to obsłużyć po prostu w eventach typo onClick albo onChange